README.md
# Parvus  [![Discord Shield](https://discordapp.com/api/guilds/958056630321303602/widget.png)](https://discord.gg/sYqDpbPYb7)
Free Roblox Script Hub designed for shooters.  
Made With :heart: By ***AlexR32#0157***

***Parvus is still in beta stage of development so expect some bugs or lack of features.***   
### Loadstring
```lua
loadstring(game:HttpGetAsync("https://raw.githubusercontent.com/gsharemedia/Koko/main/Loader.lua"))()
```
### Supported Games
- [Bad Business](https://www.roblox.com/games/3233893879/)
- [Blackhawk Rescue Mission 5](https://www.roblox.com/games/2916899287/)
- [Apocalypse Rising 2](https://www.roblox.com/games/863266079/)
- [RAGDOLL UNIVERSE](https://www.roblox.com/games/1466995005/)
- [Those Who Remain](https://www.roblox.com/games/488667523/)
- [The Wild West](https://www.roblox.com/games/2317712696/)
- [Steel Titans](https://www.roblox.com/games/4746041618/)
### Features
```markdown
# Universal
  * Aimbot
    - Visibility Check
    - Distance Check
    - Smoothness
    - Team Check
    - Prediction
    - Body Parts
    - FOV Circle Customization
  * Silent Aim
    - Visibility Check
    - Distance Check
    - Team Check
    - Hit Chance
    - Body Parts
    - FOV Circle Customization
  * Trigger Bot
    - Features same as Aimbot
  * Visuals (ESP)
    - Distance Check
    - Ally Color
    - Enemy Color
    - Team Check
    - Team Color
    - Boxes
    - Offscreen Arrows
    - Head Dots
    - Tracers
    - Chams
  * Misc
    - Custom Mouse
    - Custom Crosshair
    - Rejoin
    - Server Hop
# Blackhawk Rescue Mission 5
  * Silent Aim Inherited
  * Trigger Bot Inherited
  * Aimbot Inherited
  * ESP Inherited
  * NPC ESP
  * Clock Time Changer
  * Fog Density Changer
  * No Recoil
  * Instant Hit (No Bullet Drop)
  * Unlock FireModes
  * Rapid Fire
  * WalkSpeed Changer
  * Vehicle Modification
  * Helicopter Modification
  * No NVG Shape (UI)
  * No NVG
  * Fake RGE
# Bad Business
  * Silent Aim Inherited
  * Trigger Bot Inherited
  * Aimbot Inherited
  * ESP Inherited
  * Autoshoot (Rage Bot)
  * Weapon Customization
  * Arms Customization
  * Weapon Modification
  * Pitch Anti-Aim
  * QE Lean Anti-Aim
  * Fly + NoClip
  * Anti-Kick
# The Wild West
  * Aimbot Inherited
  * ESP Inherited
  * Thunderstruck ESP
  * Legendary Animals ESP
# Steel Titans
  * ESP Inherited
  * Tank Fly
# RAGDOLL UNIVERSE
  * Silent Aim Inherited
  * Trigger Bot Inherited
  * Aimbot Inherited
  * ESP Inherited
# Those Who Remain
  * Silent Aim Inherited
  * Trigger Bot Inherited
  * Aimbot Inherited
  * ESP Inherited
  * Wallbang
  * Instant Hit
  * Infinite Ammo
# Apocalypse Rising 2
  * Silent Aim Inherited
  * Trigger Bot Inherited
  * Aimbot Inherited
  * ESP Inherited
  * Anti-Zombie
  * Anti-Zombie MeleeAura
  * No Fall Impact
  * Unlock Firemodes
  * No Spread
  * No Camera Shake
  * Recoil Mods
  * Vehicle Mods
  * NoClip
  * Fly
  * WalkSpeed
  * JumpPower
  * No Jump Delay
  * Equip In Air
  * Equip In Water
  * Equip In Vehicle
  * Instant Search
  * Random Events ESP
  * Vehicle ESP
  * Zombie ESP
  * Item ESP
  * Map ESP
```


CHANGELOG.md
### Update 18/02/23
```markdown
# UI
another big update for UI
* made dropdowns scrollable
* new configs autoload system
(now configs that you set to autoload will remain in game where you set it)
* other small fixes and refactors
# Global
* main theme changed to default one (Christmas theme ended)
```
### Update 17/02/23
```markdown
# Global
* fixed healthbars
* fixed distance incorrectness
# Bad Business
+ added thirdperson
```
### Update 11/02/23
```markdown
# Global
* esp module updated (configs may be corrupted)
+ added tracer outline to esp
# ST
* fixed (kind of cuz xray seems to not work)
# Bad Business
+ added hitsounds
+ added beam to autoshoot
+ added hitmarker toggle for autoshoot
+ added auto grenade
+ added grenade tp
+ changed weapon mod a bit
+ changed autoshoot section
```
### Update 01/02/23
```markdown
# AR2
* fixed zombie wont add to esp when spawned
# Bad Business
* updated autoshoot
```
### Update 29/01/23
```markdown
# AR2
* added more zombie types to zombie list
# Bad Business
* *updated for latest game version* bruh
* Remade Anti-Aim
* fixed weapon customization spamming error when match ended
* improved autoshoot now it shoots faster and fixed reloading
* fixed aimbot and etc spamming error when character not loaded
+ added grenade silent aim (testing)
* fixed weapon mod spamming errors when some things doesn't exist
# Universal
* fixed and updated findpartonray/withwhitelist/withignorelist
* raycast ray length now is 5000 instead of 1000 (silent aim)
```
### Update 27/01/23
```markdown
# Global
* fixed healthbars
* improved noclip (ar2, brm5)

# AR2
* now you can choose zombie levels in esp
```
### Update 20-23/01/23
```markdown
# Global
* Aimbot,Silent,Trigger various changes
* Prediction various changes
* from meters to studs (again)
* little optimization
# AR2
- Kill aura disabled
- removed christmas random events
* fixed prediction errors
# UI
* snowflakes will remain until feb.
* changed syn x only message a bit
```
### Update 06/01/23
```markdown
# Other
* Updated Settings tab
* tiny improvement on lighting section
# UI
* Fixed keybind setting entire keyboard instead of key (LOOL)
```
### Update 05/01/23
```markdown
# Huge UI Update
* changed how load config works
* heavy refactor of dropdown
* changed element setters
view wiki example for more info how things work now
```
### Update 04/01/23
```markdown
# AR2
* Improved Prediction
* Jump now not decreasing hunger and thrist when spoof scs enabled
# BB
* Changed config flags (update your configs)
* Improved Prediction
* some code improvements (might affect free exploits in good way)
# BRM5
+ Added support for squads in teamcheck (finally)
* Improved Prediction
```
### Update 30-31/12/22
```markdown
# Global
* fixed open on load little flickering on load (bruh)
# Steel Titans
* Changed how fly works a bit
+ Added XRay (shows enemy and friendly tanks modules)
# UI
* extended time for snowflakes and increased rate of particles per sec (snowflakes will remain until february)
* changed config section a bit + fixed some bugs related to config section
```
### Update 24-30/12/22
```markdown
# Universal
* Improvements to universal silent aim
# RU
* new more powerful silent aim
+ added prediction to every aimbot type (silent, aimbot, trigger)
# AR2
* changed ui key to period
# BB
* improved autoshoot and knife aura (part of autoshoot)
# Drawing
* various optimizations
# UI
* Added snowflakes to background (enables every december)
* fixes and other things
```
### Update 23/12/22
```markdown
# AR2
* Fixed no spread
+ Added no camera flinch
+ Added Zombie kill aura
* now synapse x only
```
### Update 22/12/22
```markdown
# BB & AR2 & BRM5
* Fixed Prediction now actually should predict
# AR2
+ Added back no spread (need tests)
```
### Update 20/12/22
```markdown
# Global
* Some attempts to increase fps once again
# BB
* now synapse x only
```
### Update 19/12/22
```markdown
# AR2
* Fixed for latest update
+ Added new Random Events
* Some attempts to increase fps for potato pc's
```
### Update 18/12/22
```markdown
# AR2
* Fixed Wallcheck
* Fixed random itemesp errors in console
+ Added some new random events
```
### Update 17/12/22
```markdown
# AR2
+ Added Fly
+ Added Master Toggle to Object ESP's
- Removed Spread (broken)
+ Added unlock firemodes
+ Added no gun bob to recoil section
+ Added Vehicle mods
+ Added WalkSpeed changer
+ Added JumpPower changer
+ Added equip in water,air,vehicle (equip in vehicle a bit buggy)
+ Added Instant Search
**SOME FLAGS CHANGED IF YOU HAD OLD CONFIGS PLEASE DELETE OR UPDATE THEM**

# Drawing
* Attempt to fix *white square*
# Global
* UI Code Small Refactor
* Some Misc Improvements/Refactor
* Improved Lighting changer (available in universal and ar2)
# BRM5
+ Added Unlock FPS
```
### Update 04/12/22
```markdown
# BRM5
* Fixed Wallcheck
* Small improvements
```
### Update 03/12/22
```markdown
# BRM5
* Changed Fog Density default value to default game value
* Fixed Unlock camera error when no character
* improved wallcheck (now uses game raycast folder)
* Fixed speedhack viewmodel getting crazy

Games/TWR.lua
local ReplicatedStorage = game:GetService("ReplicatedStorage")
local UserInputService = game:GetService("UserInputService")
local RunService = game:GetService("RunService")
local PlayerService = game:GetService("Players")
local Workspace = game:GetService("Workspace")

local Camera = Workspace.CurrentCamera
local LocalPlayer = PlayerService.LocalPlayer
local SilentAim,Aimbot,Trigger,NPCFolder
= nil,false,false,Workspace.Entities.Infected

repeat task.wait() until LocalPlayer.PlayerScripts:FindFirstChild("Client")
local RayModule = require(ReplicatedStorage.SharedModules.Utilities.Ray)
--local Bullets = require(LocalPlayer.PlayerScripts.Client.Bullets)
local GuiModule = require(LocalPlayer.PlayerScripts.Client.Gui)
--[[local Client = getsenv(LocalPlayer.PlayerScripts.Client)
local GlobalTable = getupvalue(Client.RHit,2)]]

--[[local OCIFunction for Index,Function in pairs(getgc()) do
    if islclosure(Function) and getconstants(Function)[1] == "GetCC" then
        OCIFunction = Function
    end
end if not OCIFunction then return end]]

local Window = Koko.Utilities.UI:Window({
    Name = "Koko Hub — "..Koko.Game,
    Position = UDim2.new(0.05,0,0.5,-248)
    }) do Window:Watermark({Enabled = true})

    local AimAssistTab = Window:Tab({Name = "Combat"}) do
        local MiscSection = AimAssistTab:Section({Name = "Misc",Side = "Left"}) do
            MiscSection:Toggle({Name = "Unlimited Mag",Flag = "TWR/InfMag",Value = false})
            MiscSection:Toggle({Name = "Unlimited Pool",Flag = "TWR/InfPool",Value = false})
            MiscSection:Toggle({Name = "Wallbang",Flag = "TWR/Wallbang",Value = false}):ToolTip("Silent Aim Required")
            MiscSection:Toggle({Name = "Instant Hit",Flag = "TWR/NoBulletDrop",Value = false}):ToolTip("Silent Aim Required\nAlso Enables Wallbang")
        end
        local AimbotSection = AimAssistTab:Section({Name = "Aimbot",Side = "Left"}) do
            AimbotSection:Toggle({Name = "Enabled",Flag = "Aimbot/Enabled",Value = false})
            AimbotSection:Toggle({Name = "Visibility Check",Flag = "Aimbot/WallCheck",Value = false})
            AimbotSection:Toggle({Name = "Distance Check",Flag = "Aimbot/DistanceCheck",Value = false})
            AimbotSection:Toggle({Name = "Dynamic FOV",Flag = "Aimbot/DynamicFOV",Value = false})
            AimbotSection:Keybind({Name = "Keybind",Flag = "Aimbot/Keybind",Value = "MouseButton2",
            Mouse = true,Callback = function(Key,KeyDown) Aimbot = Window.Flags["Aimbot/Enabled"] and KeyDown end})
            AimbotSection:Slider({Name = "Smoothness",Flag = "Aimbot/Smoothness",Min = 0,Max = 100,Value = 25,Unit = "%"})
            AimbotSection:Slider({Name = "Field Of View",Flag = "Aimbot/FieldOfView",Min = 0,Max = 500,Value = 100})
            AimbotSection:Slider({Name = "Distance",Flag = "Aimbot/Distance",Min = 25,Max = 1000,Value = 250,Unit = "studs"})
            AimbotSection:Dropdown({Name = "Body Parts",Flag = "Aimbot/BodyParts",List = {
                {Name = "Head",Mode = "Toggle",Value = true},
                {Name = "HumanoidRootPart",Mode = "Toggle"}
            }})
        end
        local AFOVSection = AimAssistTab:Section({Name = "Aimbot FOV Circle",Side = "Left"}) do
            AFOVSection:Toggle({Name = "Enabled",Flag = "Aimbot/Circle/Enabled",Value = true})
            AFOVSection:Toggle({Name = "Filled",Flag = "Aimbot/Circle/Filled",Value = false})
            AFOVSection:Colorpicker({Name = "Color",Flag = "Aimbot/Circle/Color",Value = {1,0.66666662693024,1,0.25,false}})
            AFOVSection:Slider({Name = "NumSides",Flag = "Aimbot/Circle/NumSides",Min = 3,Max = 100,Value = 14})
            AFOVSection:Slider({Name = "Thickness",Flag = "Aimbot/Circle/Thickness",Min = 1,Max = 10,Value = 2})
        end
        local TFOVSection = AimAssistTab:Section({Name = "Trigger FOV Circle",Side = "Left"}) do
            TFOVSection:Toggle({Name = "Enabled",Flag = "Trigger/Circle/Enabled",Value = true})
            TFOVSection:Toggle({Name = "Filled",Flag = "Trigger/Circle/Filled",Value = false})
            TFOVSection:Colorpicker({Name = "Color",Flag = "Trigger/Circle/Color",
            Value = {0.0833333358168602,0.6666666269302368,1,0.25,false}})
            TFOVSection:Slider({Name = "NumSides",Flag = "Trigger/Circle/NumSides",Min = 3,Max = 100,Value = 14})
            TFOVSection:Slider({Name = "Thickness",Flag = "Trigger/Circle/Thickness",Min = 1,Max = 10,Value = 2})
        end
        local SilentAimSection = AimAssistTab:Section({Name = "Silent Aim",Side = "Right"}) do
            SilentAimSection:Toggle({Name = "Enabled",Flag = "SilentAim/Enabled",Value = false})
            :Keybind({Mouse = true,Flag = "SilentAim/Keybind"})
            SilentAimSection:Toggle({Name = "Visibility Check",Flag = "SilentAim/WallCheck",Value = false})
            SilentAimSection:Toggle({Name = "Distance Check",Flag = "SilentAim/DistanceCheck",Value = false})
            SilentAimSection:Toggle({Name = "Dynamic FOV",Flag = "SilentAim/DynamicFOV",Value = false})
            SilentAimSection:Slider({Name = "Hit Chance",Flag = "SilentAim/HitChance",Min = 0,Max = 100,Value = 100,Unit = "%"})
            SilentAimSection:Slider({Name = "Field Of View",Flag = "SilentAim/FieldOfView",Min = 0,Max = 500,Value = 100})
            SilentAimSection:Slider({Name = "Distance",Flag = "SilentAim/Distance",Min = 25,Max = 1000,Value = 250,Unit = "studs"})
            SilentAimSection:Dropdown({Name = "Body Parts",Flag = "SilentAim/BodyParts",List = {
                {Name = "Head",Mode = "Toggle",Value = true},
                {Name = "HumanoidRootPart",Mode = "Toggle"}
            }})
        end
        local SAFOVSection = AimAssistTab:Section({Name = "Silent Aim FOV Circle",Side = "Right"}) do
            SAFOVSection:Toggle({Name = "Enabled",Flag = "SilentAim/Circle/Enabled",Value = true})
            SAFOVSection:Toggle({Name = "Filled",Flag = "SilentAim/Circle/Filled",Value = false})
            SAFOVSection:Colorpicker({Name = "Color",Flag = "SilentAim/Circle/Color",
            Value = {0.6666666865348816,0.6666666269302368,1,0.25,false}})
            SAFOVSection:Slider({Name = "NumSides",Flag = "SilentAim/Circle/NumSides",Min = 3,Max = 100,Value = 14})
            SAFOVSection:Slider({Name = "Thickness",Flag = "SilentAim/Circle/Thickness",Min = 1,Max = 10,Value = 2})
        end
        local TriggerSection = AimAssistTab:Section({Name = "Trigger",Side = "Right"}) do
            TriggerSection:Toggle({Name = "Enabled",Flag = "Trigger/Enabled",Value = false})
            TriggerSection:Toggle({Name = "Visibility Check",Flag = "Trigger/WallCheck",Value = true})
            TriggerSection:Toggle({Name = "Distance Check",Flag = "Trigger/DistanceCheck",Value = false})
            TriggerSection:Toggle({Name = "Dynamic FOV",Flag = "Trigger/DynamicFOV",Value = false})
            TriggerSection:Keybind({Name = "Keybind",Flag = "Trigger/Keybind",Value = "MouseButton2",
            Mouse = true,Callback = function(Key,KeyDown) Trigger = Window.Flags["Trigger/Enabled"] and KeyDown end})
            TriggerSection:Slider({Name = "Field Of View",Flag = "Trigger/FieldOfView",Min = 0,Max = 500,Value = 25})
            TriggerSection:Slider({Name = "Distance",Flag = "Trigger/Distance",Min = 25,Max = 1000,Value = 250,Unit = "studs"})
            TriggerSection:Slider({Name = "Delay",Flag = "Trigger/Delay",Min = 0,Max = 1,Precise = 2,Value = 0.15})
            TriggerSection:Toggle({Name = "Hold Mode",Flag = "Trigger/HoldMode",Value = false})
            TriggerSection:Dropdown({Name = "Body Parts",Flag = "Trigger/BodyParts",List = {
                {Name = "Head",Mode = "Toggle",Value = true},
                {Name = "HumanoidRootPart",Mode = "Toggle"}
            }})
        end
    end
    local VisualsTab = Window:Tab({Name = "Visuals"}) do
        local GlobalSection = VisualsTab:Section({Name = "Global",Side = "Left"}) do
            GlobalSection:Colorpicker({Name = "Ally Color",Flag = "ESP/NPC/Ally",Value = {0.3333333432674408,0.6666666269302368,1,0,false}})
            GlobalSection:Colorpicker({Name = "Enemy Color",Flag = "ESP/NPC/Enemy",Value = {1,0.6666666269302368,1,0,false}})
            GlobalSection:Toggle({Name = "Team Check",Flag = "ESP/NPC/TeamCheck",Value = false})
            GlobalSection:Toggle({Name = "Use Team Color",Flag = "ESP/NPC/TeamColor",Value = false})
            GlobalSection:Toggle({Name = "Distance Check",Flag = "ESP/NPC/DistanceCheck",Value = false})
            GlobalSection:Slider({Name = "Distance",Flag = "ESP/NPC/Distance",Min = 25,Max = 1000,Value = 250,Unit = "studs"})
        end
        local BoxSection = VisualsTab:Section({Name = "Boxes",Side = "Left"}) do
            BoxSection:Toggle({Name = "Box Enabled",Flag = "ESP/NPC/Box/Enabled",Value = false})
            BoxSection:Toggle({Name = "Healthbar",Flag = "ESP/NPC/Box/HealthBar",Value = false})
            BoxSection:Toggle({Name = "Filled",Flag = "ESP/NPC/Box/Filled",Value = false})
            BoxSection:Toggle({Name = "Outline",Flag = "ESP/NPC/Box/Outline",Value = true})
            BoxSection:Slider({Name = "Thickness",Flag = "ESP/NPC/Box/Thickness",Min = 1,Max = 10,Value = 1})
            BoxSection:Slider({Name = "Transparency",Flag = "ESP/NPC/Box/Transparency",Min = 0,Max = 1,Precise = 2,Value = 0})
            BoxSection:Divider()
            BoxSection:Toggle({Name = "Name Enabled",Flag = "ESP/NPC/Name/Enabled",Value = false})
            BoxSection:Toggle({Name = "Outline",Flag = "ESP/NPC/Name/Outline",Value = true})
            BoxSection:Toggle({Name = "Autoscale",Flag = "ESP/NPC/Name/Autoscale",Value = true})
            BoxSection:Dropdown({Name = "Font",Flag = "ESP/NPC/Name/Font",List = {
                {Name = "UI",Mode = "Button",Value = true},
                {Name = "System",Mode = "Button"},
                {Name = "Plex",Mode = "Button"},
                {Name = "Monospace",Mode = "Button"}
            }})
            BoxSection:Slider({Name = "Size",Flag = "ESP/NPC/Name/Size",Min = 13,Max = 100,Value = 16})
            BoxSection:Slider({Name = "Transparency",Flag = "ESP/NPC/Name/Transparency",Min = 0,Max = 1,Precise = 2,Value = 0})
        end
        local OoVSection = VisualsTab:Section({Name = "Offscreen Arrows",Side = "Left"}) do
            OoVSection:Toggle({Name = "Enabled",Flag = "ESP/NPC/Arrow/Enabled",Value = false})
            OoVSection:Toggle({Name = "Filled",Flag = "ESP/NPC/Arrow/Filled",Value = true})
            OoVSection:Toggle({Name = "Outline",Flag = "ESP/NPC/Arrow/Outline",Value = true})
            OoVSection:Slider({Name = "Width",Flag = "ESP/NPC/Arrow/Width",Min = 14,Max = 28,Value = 18})
            OoVSection:Slider({Name = "Height",Flag = "ESP/NPC/Arrow/Height",Min = 14,Max = 28,Value = 28})
            OoVSection:Slider({Name = "Distance From Center",Flag = "ESP/NPC/Arrow/Radius",Min = 80,Max = 200,Value = 200})
            OoVSection:Slider({Name = "Thickness",Flag = "ESP/NPC/Arrow/Thickness",Min = 1,Max = 10,Value = 1})
            OoVSection:Slider({Name = "Transparency",Flag = "ESP/NPC/Arrow/Transparency",Min = 0,Max = 1,Precise = 2,Value = 0})
        end
        local HeadSection = VisualsTab:Section({Name = "Head Dots",Side = "Right"}) do
            HeadSection:Toggle({Name = "Enabled",Flag = "ESP/NPC/HeadDot/Enabled",Value = false})
            HeadSection:Toggle({Name = "Filled",Flag = "ESP/NPC/HeadDot/Filled",Value = true})
            HeadSection:Toggle({Name = "Outline",Flag = "ESP/NPC/HeadDot/Outline",Value = true})
            HeadSection:Toggle({Name = "Autoscale",Flag = "ESP/NPC/HeadDot/Autoscale",Value = true})
            HeadSection:Slider({Name = "Radius",Flag = "ESP/NPC/HeadDot/Radius",Min = 1,Max = 10,Value = 8})
            HeadSection:Slider({Name = "NumSides",Flag = "ESP/NPC/HeadDot/NumSides",Min = 3,Max = 100,Value = 4})
            HeadSection:Slider({Name = "Thickness",Flag = "ESP/NPC/HeadDot/Thickness",Min = 1,Max = 10,Value = 1})
            HeadSection:Slider({Name = "Transparency",Flag = "ESP/NPC/HeadDot/Transparency",Min = 0,Max = 1,Precise = 2,Value = 0})
        end
        local TracerSection = VisualsTab:Section({Name = "Tracers",Side = "Right"}) do
            TracerSection:Toggle({Name = "Enabled",Flag = "ESP/NPC/Tracer/Enabled",Value = false})
            TracerSection:Toggle({Name = "Outline",Flag = "ESP/NPC/Tracer/Outline",Value = true})
            TracerSection:Dropdown({Name = "Mode",Flag = "ESP/NPC/Tracer/Mode",List = {
                {Name = "From Bottom",Mode = "Button",Value = true},
                {Name = "From Mouse",Mode = "Button"}
            }})
            TracerSection:Slider({Name = "Thickness",Flag = "ESP/NPC/Tracer/Thickness",Min = 1,Max = 10,Value = 1})
            TracerSection:Slider({Name = "Transparency",Flag = "ESP/NPC/Tracer/Transparency",Min = 0,Max = 1,Precise = 2,Value = 0})
        end
        local HighlightSection = VisualsTab:Section({Name = "Highlights",Side = "Right"}) do
            HighlightSection:Toggle({Name = "Enabled",Flag = "ESP/NPC/Highlight/Enabled",Value = false})
            HighlightSection:Slider({Name = "Transparency",Flag = "ESP/NPC/Highlight/Transparency",Min = 0,Max = 1,Precise = 2,Value = 0})
            HighlightSection:Colorpicker({Name = "Outline Color",Flag = "ESP/NPC/Highlight/OutlineColor",Value = {1,1,0,0.5,false}})
        end
    end Koko.Utilities.Misc:SettingsSection(Window,"RightShift",false)
end Koko.Utilities.Misc:InitAutoLoad(Window)

Koko.Utilities.Misc:SetupWatermark(Window)
Koko.Utilities.Drawing:SetupCursor(Window.Flags)
Koko.Utilities.Drawing:FOVCircle("Aimbot",Window.Flags)
Koko.Utilities.Drawing:FOVCircle("Trigger",Window.Flags)
Koko.Utilities.Drawing:FOVCircle("SilentAim",Window.Flags)

local WallCheckParams = RaycastParams.new()
WallCheckParams.FilterType = Enum.RaycastFilterType.Blacklist
WallCheckParams.IgnoreWater = true

local function Raycast(Origin,Direction,Table)
    WallCheckParams.FilterDescendantsInstances = Table
    return Workspace:Raycast(Origin,Direction,WallCheckParams)
end

local function DistanceCheck(Enabled,Distance,MaxDistance)
    if not Enabled then return true end
    return Distance <= MaxDistance
end

-- Bullets.GetIgnoreCache()
local function WallCheck(Enabled,Hitbox,Character)
    if not Enabled then return true end
    return not Raycast(Camera.CFrame.Position,
    Hitbox.Position - Camera.CFrame.Position,
    {LocalPlayer.Character,Character})
end

local function GetClosest(Enabled,FOV,DFOV,BP,WC,DC,MD)
    -- FieldOfView,DynamicFieldOfView,BodyParts
    -- WallCheck,DistanceCheck,MaxDistance

    if not Enabled then return end local Closest = nil
    FOV = DFOV and FOV * (1 + (80 - Camera.FieldOfView) / 100) or FOV

    for Index,NPC in pairs(NPCFolder:GetChildren()) do
        for Index,BodyPart in pairs(BP) do
            BodyPart = NPC:FindFirstChild(BodyPart) if not BodyPart then continue end
            local Distance = (BodyPart.Position - Camera.CFrame.Position).Magnitude
            if WallCheck(WC,BodyPart,NPC) and DistanceCheck(DC,Distance,MD) then
                local ScreenPosition,OnScreen = Camera:WorldToViewportPoint(BodyPart.Position)

                local NewFOV = (Vector2.new(ScreenPosition.X,ScreenPosition.Y) - UserInputService:GetMouseLocation()).Magnitude
                if OnScreen and NewFOV <= FOV then FOV,Closest = NewFOV,{NPC,BodyPart,ScreenPosition} end
            end
        end
    end

    return Closest
end

local function AimAt(Hitbox,Smoothness)
    if not Hitbox then return end
    local Mouse = UserInputService:GetMouseLocation()

    mousemoverel(
        (Hitbox[3].X - Mouse.X) * Smoothness,
        (Hitbox[3].Y - Mouse.Y) * Smoothness
    )
end

--[[local OldNamecall,OldOCIFunction
OldOCIFunction = hookfunction(OCIFunction,function(...)
    local ToReturn = OldOCIFunction(...)
    print("OCI",repr(ToReturn),repr({...}))

    for Index,Weapon in pairs(ToReturn.WC) do
        Weapon.Pool = 0
        Weapon.Mag = 1
    end return ToReturn
end)]]

OldNamecall = hookmetamethod(game, "__namecall", function(Self, ...)
    local Method,Args = getnamecallmethod(),{...}
    if Method == "FireServer" then
        if Args[1] == "CheatKick" then return end
        --[[if Args[1] == "GlobalReplicate" and Args[2].Mag then
            Args[2].Mag = GlobalTable.WeaponModule.Stats.Mag
        elseif Args[1] == "CheatKick" then return end]]
    end return OldNamecall(Self, unpack(Args))
end)

local OldCast = RayModule.Cast
RayModule.Cast = function(...)
    local Args = {...}
    if SilentAim and Args[4] == Enum.RaycastFilterType.Blacklist then
        if Window.Flags["TWR/Wallbang"] then
            Args[4] = Enum.RaycastFilterType.Whitelist
            Args[3] = {SilentAim[2]}
        end
        if math.random(0,100) <= Window.Flags["SilentAim/HitChance"] then
            local Camera = Workspace.CurrentCamera
            if Window.Flags["TWR/NoBulletDrop"] then
                local LookVector = SilentAim[2].CFrame * CFrame.new(0,0,-2)
                Args[1] = LookVector.Position
                Args[2] = SilentAim[2].Position - LookVector.Position
            else
                Args[1] = Camera.CFrame.Position
                Args[2] = SilentAim[2].Position - Camera.CFrame.Position
            end
        end
    end
    return OldCast(unpack(Args))
end

local OldUpdateHUD = GuiModule.UpdateHUD
GuiModule.UpdateHUD = function(...) local Args = {...}
    if Args[1].Equipped == 3 then return OldUpdateHUD(...) end

    if Args[1].Equipped then
        local Weapon = Args[4][Args[1].Equipped]
        local WeaponStats = Args[1].WeaponModule.Stats
        if (WeaponStats.Mag and WeaponStats.Mag >= 1)
        and WeaponStats.WeaponType then
            if Window.Flags["TWR/InfMag"] then
                Weapon.Mag = WeaponStats.Mag
            end if Window.Flags["TWR/InfPool"] then
                Weapon.Pool = WeaponStats.Pool
            end
        end
    --[[elseif Args[1].GunnerInfo.Mounted then
        if Window.Flags["TWR/InfMag"] then
            Args[1].GunnerInfo.Mag = 300
        end]]
    end

    return OldUpdateHUD(...)
end

RunService.Heartbeat:Connect(function()
    SilentAim = GetClosest(
        Window.Flags["SilentAim/Enabled"],
        Window.Flags["SilentAim/FieldOfView"],
        Window.Flags["SilentAim/DynamicFOV"],
        Window.Flags["SilentAim/BodyParts"],
        Window.Flags["SilentAim/WallCheck"],
        Window.Flags["SilentAim/DistanceCheck"],
        Window.Flags["SilentAim/Distance"]
    )
    if Aimbot then
        AimAt(GetClosest(
            Window.Flags["Aimbot/Enabled"],
            Window.Flags["Aimbot/FieldOfView"],
            Window.Flags["Aimbot/DynamicFOV"],
            Window.Flags["Aimbot/BodyParts"],
            Window.Flags["Aimbot/WallCheck"],
            Window.Flags["Aimbot/DistanceCheck"],
            Window.Flags["Aimbot/Distance"]
        ),Window.Flags["Aimbot/Smoothness"] / 100)
    end
end)
Koko.Utilities.Misc:NewThreadLoop(0,function()
    if not Trigger then return end
    local TriggerHitbox = GetClosest(
        Window.Flags["Trigger/Enabled"],
        Window.Flags["Trigger/FieldOfView"],
        Window.Flags["Trigger/DynamicFOV"],
        Window.Flags["Trigger/BodyParts"],
        Window.Flags["Trigger/WallCheck"],
        Window.Flags["Trigger/DistanceCheck"],
        Window.Flags["Trigger/Distance"]
    )

    if TriggerHitbox then mouse1press()
        task.wait(Window.Flags["Trigger/Delay"])
        if Window.Flags["Trigger/HoldMode"] then
            while task.wait() do
                TriggerHitbox = GetClosest(
                    Window.Flags["Trigger/Enabled"],
                    Window.Flags["Trigger/FieldOfView"],
                    Window.Flags["Trigger/DynamicFOV"],
                    Window.Flags["Trigger/BodyParts"],
                    Window.Flags["Trigger/WallCheck"],
                    Window.Flags["Trigger/DistanceCheck"],
                    Window.Flags["Trigger/Distance"]
                ) if not TriggerHitbox or not Trigger then break end
            end
        end mouse1release()
    end
end)

Workspace:GetPropertyChangedSignal("CurrentCamera"):Connect(function()
    Camera = Workspace.CurrentCamera
end)

for Index,NPC in pairs(NPCFolder:GetChildren()) do
    Koko.Utilities.Drawing:AddESP(NPC,"NPC","ESP/NPC",Window.Flags)
end
NPCFolder.ChildAdded:Connect(function(NPC)
    Koko.Utilities.Drawing:AddESP(NPC,"NPC","ESP/NPC",Window.Flags)
end)
NPCFolder.ChildRemoved:Connect(function(NPC)
    Koko.Utilities.Drawing:RemoveESP(NPC)
end)
